### ISSUE TITLE
May DevOps Infrastructure and Automation (3/3)

### ISSUE THREAD

**Issue by elisabettai** [ITISFoundation/osparc-issues#1102]
Description:
Develop DevOps infrastructure empowering users to perform more monitoring and troubleshooting tasks themselves, thus facilitating debugging and reducing the required SIM-Core manpower in the longer run.
Foreseen functionality includes:
 - Provide increased automation for the following routine DevOps tasks:
    - Better organization of docker images in registry (validation of metadata, deletion of faulty images)
    - Simplification of interaction with database to have command line tools for:
       - bulk add/remove/upgrade access rights to services for a user/group
       - bulk remove/add/upgrade services from projects 
       - automate resource limits assignment for users based on service and user/group-id (also relates to API extensions & computational backend refactoring discussed above)
    - Information on which users have currently which studies open
    - Development of tools to collect resource usage (compute & storage) information for potential billing (part of sustainability plans) based on
       - services
       - user groups
    - Housekeeping: Detecting stale (old/unused) user accounts and proactively find corrupt projects/services and unused data for clean-up by DevOps

- Web-Application Monitoring: Automatic collection of metrics upon any unexpected failures and nascent bugs, which will be used to assess the user-experience during our development cycles. This will help to assess if user experience degrades or improves and how much time for maintenance is required flexibly. There is no need to rely only on traditional, mostly slow feedback cycles (such as dissatisfied users’ email feedback)

Deliverable: 
Extended DevOps infrastructure: housekeeping, simplification of interaction with database, web-application monitoring

Acceptance criteria: 
Infrastructure is in place that provides increased DevOps automation and web-application monitoring

Out of scope: –

Deadline:
Q1

[wrike Y7-MS 7.1.3: DevOps Infrastructure and Automation (3/3)](https://www.wrike.com/open.htm?id=1158230087)

**Comment by SCA-ZMT** [comment]
Prepare for review.

**Comment by elisabettai** [comment]
I am writing this up for the NIH reporting.

I am sure there's a lot we're done for "housekeeping, simplification of interaction with database, web-application monitoring"

We can for sure mention the new dashboards on Graphana to monitor running machines, right @YuryHrytsuk @sanderegg? Do you have more details that we can share?

Something about interaction with the DB? Maybe we can mention scripts or other tools that we used to migrate TIP to AWS? @matusdrobuliak66 

Anything else for housekeeping @mrnicegyu11?

**Comment by elisabettai** [comment]
In the reporting for Q3, we considered this as complete with the following justification:

_New monitoring infrastructure has been added, that allows to visualize the usage of AWS machines created “on demand” via the autoscaling microservice. The data can be visualized in real-time in dashboards powered by Grafana. In this way, it can readily be visualized how many machines are running, for how long, etc…, e.g., to facilitate cost optimization. The functionality also allows to raise automatic alerts, e.g., when machines have been idling for a prolonged period, without being properly terminated.
A series of scripts have been developed to facilitate interaction with the database. They were used during the reporting period to migrate a full o2S2PARC deployment (users, studies, and generated data; see also [Task 9c Operations](https://github.com/ITISFoundation/osparc-issues/issues/1109). In addition, recent platform security improvements have complicated the operation of remote-connecting to machines, hampering DevOps, such that facilitating scripts had to be developed.
To streamline platform housekeeping, we improved the management of user-data backups, also reducing costs without compromising data safety. For cost management reasons, links from incomplete data uploads now expire after a specified period._
